% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/plotQC.R
\name{plotQC}
\alias{plotQC}
\title{plotQC}
\usage{
plotQC(
  spe,
  type = c("hist", "scatter", "spots"),
  x_coord = NULL,
  y_coord = NULL,
  in_tissue = "in_tissue",
  metric_x = "cell_count",
  metric_y = "sum",
  discard = "discard",
  nbins = 100,
  threshold_x = NULL,
  threshold_y = NULL,
  trend = TRUE,
  marginal = TRUE,
  y_reverse = TRUE
)
}
\arguments{
\item{spe}{(SpatialExperiment) Input data, assumed to be a
\code{SpatialExperiment} object.}

\item{type}{(character) Type of QC plot. Options are "hist", "scatter", and
"spots". See details in description.}

\item{x_coord}{(character) Name of column in \code{spatialCoords} containing
x-coordinates. Default = NULL, which selects the first column. Used for
spot-based plots.}

\item{y_coord}{(character) Name of column in \code{spatialCoords} containing
y-coordinates. Default = NULL, which selects the second column. Used for
spot-based plots.}

\item{in_tissue}{(character) Name of column in \code{colData} identifying
spots over tissue, e.g. "in_tissue" for 10x Genomics Visium data. If this
argument is provided, only spots over tissue will be shown. Alternatively,
set to NULL to display all spots. Default = "in_tissue".}

\item{metric_x}{(character) Name of column in \code{colData} containing QC
metric to plot on x-axis (e.g. "cell_count" for number of cells per spot).
Default = "cell_count". Required for histogram and scatterplots.}

\item{metric_y}{(character) Name of column in \code{colData} containing QC
metric to plot on y-axis (e.g. "sum" for number of detected transcripts, or
"high_mito" for binary ). Default = "sum". Required for
histogram and scatterplots.}

\item{discard}{(character) Name of column in \code{colData} identifying
discarded spots that do not meet filtering thresholds, which will be
highlighted on a spot-based plot. Default = "discard". Optional for
spot-based plots.}

\item{nbins}{(numeric) Adjusting the histogram bin width. Optional for
spot-based and scatterplots.}

\item{threshold_x}{(numeric) Filtering threshold for QC metric on x-axis,
which will be highlighted with a vertical bar. Default = NULL. Optional for
scatterplots.}

\item{threshold_y}{(numeric) Filtering threshold for QC metric on y-axis,
which will be highlighted with a horizontal bar. Default = NULL. Optional
for scatterplots.}

\item{trend}{(logical) Whether to display a smoothed trend (loess) for
scatterplots. Default = TRUE. Optional for scatterplots.}

\item{marginal}{(logical) Whether to display marginal histograms for
scatterplots. Default = TRUE. Optional for scatterplots.}

\item{y_reverse}{(logical) Whether to reverse y coordinates, which is often
required for 10x Genomics Visium data. Default = TRUE.}
}
\value{
Returns a ggplot object. Additional plot elements can be added as
  ggplot elements (e.g. title, labels, formatting, etc).
}
\description{
Quality control (QC) plots for spatially resolved transcriptomics data.
}
\details{
Function to generate plots for quality control (QC) purposes for spatially
resolved transcriptomics data.

The following types of QC plots are available:

- Histogram (\code{type} = "bar") for a single QC metric, e.g. number of counts
per spot. For number of counts per spot, the histogram highlights spots with
derived flags, e.g. low library size.
- Scatterplot (\code{type} = "scatter") comparing two QC metrics, e.g. number
of detected features vs. number of cells per spot, with optional vertical and
horizontal lines highlighting QC filtering thresholds.
- Spots (\code{type} = "spots") i.e. spots in spatial (x-y) coordinates,
highlighting discarded spots that do not meet filtering thresholds.
}
\examples{
library(STexampleData)
spe <- Visium_humanDLPFC()
plotQC(spe, type = "hist", metric_x = "cell_count")
colData(spe)$sum <- colSums(counts(spe))
plotQC(spe, type = "scatter", metric_x = "cell_count", metric_y = "sum")

}
